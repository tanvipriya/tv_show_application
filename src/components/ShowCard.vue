<template>
  <div class="show-card">
    <img :src="show.image?.medium || placeholder" :alt="show.name" />
    <h3 class="show-name">{{ show.name }}</h3>
    <p class="show-rating">‚≠ê {{ show.rating?.average || "N/A" }}</p>
    <p class="show-year">{{ show.premiered?.slice(0, 4) || "N/A" }}</p>
  </div>
</template>

<script setup>
defineProps({
  show: Object,
});

const placeholder = "https://via.placeholder.com/150x225?text=No+Image";
</script>

<style scoped>
.show-card {
  min-width: 150px;
  margin: 0 8px;
  text-align: center;
  flex-shrink: 0;
}
.show-card img {
  width: 100%;
  border-radius: 8px;
  object-fit: cover;
}
.show-name {
  font-size: 0.9rem;
  margin: 4px 0 2px 0;
}
.show-rating,
.show-year {
  font-size: 0.8rem;
  margin: 1px 0;
}
</style>
