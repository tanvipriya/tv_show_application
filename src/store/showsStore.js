import { defineStore } from 'pinia';
import axios from 'axios';

export const useShowsStore = defineStore('shows', {
    state: () => ({
        showsByGenre: {}, // { drama: [...], comedy: [...] }
        loading: false,
        error: null
    }),
    actions: {
        async fetchShowsByGenre(genre) {
            this.loading = true;
            this.error = null;
            try {
                // Fetch first 2 pages (TVMaze has 250 shows per page)
                const pages = [0, 1];
                let allShows = [];

                for (const page of pages) {
                    const res = await axios.get(`https://api.tvmaze.com/shows?page=${page}`);
                    allShows = allShows.concat(res.data);
                }

                // Filter by genre
                const filtered = allShows.filter((show) => show.genres.includes(genre));

                // Sort by rating descending (null rating goes last)
                filtered.sort((a, b) => (b.rating.average || 0) - (a.rating.average || 0));

                this.showsByGenre[genre] = filtered;
            } catch (err) {
                this.error = err.message || 'Failed to fetch shows';
            } finally {
                this.loading = false;
            }
        }
    }
});
