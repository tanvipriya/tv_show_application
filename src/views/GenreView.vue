<template>
  <div class="genre-view">
    <h1>{{ genre }} Shows</h1>

    <div class="shows-grid" ref="scrollContainer">
      <ShowCard v-for="show in shows" :key="show.id" :show="show" />
    </div>

    <p v-if="loading" class="loading">Loading...</p>
    <p v-if="error" class="error">{{ error }}</p>
  </div>
</template>

<script setup>
import { ref, computed, onMounted } from "vue";
import { useRoute } from "vue-router";
import { useShowsStore } from "@/store/showsStore";
import ShowCard from "@/components/ShowCard.vue";

const route = useRoute();
const genre = route.params.genre;

const store = useShowsStore();
const loading = computed(() => store.loading);
const error = computed(() => store.error);

const shows = computed(() => store.showsByGenre[genre] || []);

// Infinite scroll logic
const scrollContainer = ref(null);

async function loadNextBatch() {
  if (!loading.value) {
    await store.fetchNextPage(genre);
  }
}

function handleScroll() {
  const container = scrollContainer.value;
  if (!container || loading.value) return;

  if (
    container.scrollTop + container.clientHeight >=
    container.scrollHeight - 50
  ) {
    loadNextBatch();
  }
}

onMounted(async () => {
  if (!store.showsByGenre[genre]) {
    await store.fetchNextPage(genre); // load first batch
  }
  scrollContainer.value?.addEventListener("scroll", handleScroll);
});
</script>

<style scoped>
.genre-view {
  padding: 16px;
}

h1 {
  font-size: 1.5rem;
  margin-bottom: 16px;
}

.shows-grid {
  display: grid;
  grid-template-columns: repeat(auto-fill, minmax(150px, 1fr));
  gap: 12px;
  max-height: 80vh; /* optional to make container scrollable */
  overflow-y: auto;
}

.loading,
.error {
  text-align: center;
  margin: 16px 0;
}
</style>
