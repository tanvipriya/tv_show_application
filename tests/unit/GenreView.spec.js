import { createRouter, createMemoryHistory } from "vue-router";
import { mount } from "@vue/test-utils";
import GenreView from "@/views/GenreView.vue";
import ShowDetail from "@/views/ShowDetailView.vue";
import ShowCard from "@/components/ShowCard.vue";
import { createTestingPinia } from "@pinia/testing";
import { useShowsStore } from "@/store/showsStore";

const mockShows = [
    { id: 1, name: "Show One", rating: 8, image: null },
    { id: 2, name: "Show Two", rating: 9, image: "https://example.com/show2.jpg" },
];

describe("GenreView.vue - stable coverage with snapshot", () => {
    let router;
    let pushSpy;
    let pinia;
    let store;

    beforeEach(async () => {
        router = createRouter({
            history: createMemoryHistory(),
            routes: [
                { path: "/", name: "Home", component: { template: "<div />" } },
                { path: "/genre/:genre", name: "GenreView", component: GenreView },
                { name: "ShowDetail", path: "/show/:id", component: ShowDetail },
            ],
        });

        pushSpy = jest.spyOn(router, "push");
        pinia = createTestingPinia({ stubActions: false });
        store = useShowsStore();

        store.loading = false;
        store.error = null;
        store.showsByGenre = { Drama: mockShows };

        await router.push("/genre/Drama");
        await router.isReady();
    });

    it("renders show cards and navigates to ShowDetail on click", async () => {
        const wrapper = mount(GenreView, {
            global: {
                plugins: [pinia, router],
                components: { ShowCard },
                mocks: { hasMore: false },
            },
        });

        await wrapper.vm.$nextTick();

        // Check correct number of show cards
        const cards = wrapper.findAllComponents(ShowCard);
        expect(cards.length).toBe(mockShows.length);

        // Trigger navigation on first card click
        await cards[0].trigger("click");
        expect(pushSpy).toHaveBeenCalledWith({
            name: "ShowDetail",
            params: { id: 1 },
        });

        // --- Snapshot Test ---
        expect(wrapper.html()).toMatchSnapshot();
    });
});
