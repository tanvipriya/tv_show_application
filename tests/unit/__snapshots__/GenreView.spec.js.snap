// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GenreView.vue - stable coverage with snapshot renders show cards and navigates to ShowDetail on click 1`] = `
<div class="genre-view">
  <!-- Back Button --><button class="back-btn">← Back to Dashboard</button>
  <h1>Drama Shows</h1>
  <div class="shows-grid"><a href="/show/1" class="show-link">
      <div class="show-card">
        <!--v-if-->
        <div class="details">
          <h3>Show One</h3>
          <p>⭐ N/A</p>
          <p>—</p>
        </div>
      </div>
    </a><a href="/show/2" class="show-link">
      <div class="show-card">
        <!--v-if-->
        <div class="details">
          <h3>Show Two</h3>
          <p>⭐ N/A</p>
          <p>—</p>
        </div>
      </div>
    </a></div>
  <!--v-if-->
  <!--v-if-->
</div>
`;
